### PR TITLE
PCVL-810-implement-post-select-interface

### PR DESCRIPTION
**Pull request by MarionQuandela**

### CODE REVIEW

File: perceval/utils/postselect.py [hunk 1]
```diff
@@ -26,19 +26,35 @@
...
 
-    :param str_repr: string representation of the selection logic. The format is: "cond_1 & cond_2 & ... & cond_n"
-        where cond_i is "[mode list] <operator> <photon count>" (supported operators are ==, > and <)
+    Init uses a string representation of the selection logic. The format is: "cond_1 OP cond_2 OP ... OP cond_n"
+    where cond_i is "[mode list] <operator> <photon count>" (supported operators are ==, >, <, >=, <=).
+
+    Logic operators between conditions can be
+        - "AND", "and", "&"
+        - "OR", "or", "|"
+        - "XOR", "xor"
```
Comment (ericbrts) [inline perceval/utils/postselect.py:53]: - "XOR", "xor", "^"

File: perceval/utils/postselect.py [hunk 2]
```diff
@@ -50,175 +66,49 @@ class PostSelect:
...
-                    condition_count += 1
-
-            except json.decoder.JSONDecodeError as e:
-                raise RuntimeError(f"Could not interpret input string '{str_repr}': {e}")
-            if condition_count != str_repr.count("&") + 1:
-                raise RuntimeError(f"Could not interpret input string '{str_repr}': Invalid format")
-
-    def eq(self, indexes, value: int):
+    @deprecated(version="0.12.0", reason="Use merge instead")
+    def eq(self, indexes: int | list[int], value: int) -> Self:
```
Comment (ericbrts) [inline perceval/utils/postselect.py:70]: I'd say remove eq, gt, lt, ge, le and add a few lines about it in legacy.rst?

File: perceval/utils/postselect.py [hunk 3]
```diff
@@ -26,19 +26,35 @@
...
+
+from perceval.utils.logging import deprecated
 
-import json
-import re
+operands: TypeAlias = LogicOperator
 
 
-class PostSelect:
+class PostSelect(ex_PostSelect):
```
Comment (ericbrts) [inline perceval/utils/postselect.py:43]: Can we drop this inheritance if we drop the methods eq, gt, etc. methods too?
That way, we'd just have a TypeAlias for PostSelect (?)

**Review by ericbrts — CHANGES_REQUESTED** [review]

File: perceval/utils/postselect.py [hunk 4]
```diff
@@ -26,199 +26,43 @@
...
-        """
-        Merge with other PostSelect. Updates the current instance.
-
-        :param other: Another PostSelect instance
-        """
-        for operator, cond in other._conditions.items():
-            for indexes, value in cond:
-                self._add_condition(indexes, operator, value)
+>>> ps = PostSelect("[0,1] == 1 & [2] > 1") # Means "I want exactly one photon in mode 0 or 1, and at least one photon in mode 2"
+>>> ps = PostSelect().eq([0,1], 1).gt(2, 1) # Same as above
```
Comment (MarionQuandela) [inline perceval/utils/postselect.py:55]: nop

File: perceval/utils/postselect.py [hunk 5]
```diff
@@ -26,206 +26,190 @@
...
+from .statevector import BSDistribution, StateVector, BasicState
 
-class PostSelect:
-    """PostSelect is a callable post-selection intended to filter out unwanted output states. It is designed to be a
-    user-friendly description of any post-selection logic.
 
-    :param str_repr: string representation of the selection logic. The format is: "cond_1 & cond_2 & ... & cond_n"
-        where cond_i is "[mode list] <operator> <photon count>" (supported operators are ==, > and <)
+class PostSelect(xq.PostSelect):
+    """PostSelect is a callable basic state predicate intended to filter out unwanted basic states. It is designed to be an
```
Comment (ericbrts) [inline perceval/utils/postselect.py:39]: a user-friendly

File: perceval/utils/postselect.py [hunk 6]
```diff
@@ -26,206 +26,190 @@
...
+            - string that represent an operator.
+            - Operator within a condition can be:
+                - Equal: "=="
+                - Greater than: ">"
+                - Greater or equal to: ">="
+                - Less than: "<"
+                - Less or equal to: "<="
+
+        - Photon count:
+            - string that represent an non negative integer
```
Comment (ericbrts) [inline perceval/utils/postselect.py:68]: a non-negative

File: perceval/utils/postselect.py [hunk 7]
```diff
@@ -26,206 +26,190 @@
...
+                    - "xor"
+                    - "^" (default serialization string representation)
+            - NOT:
+                - can be used in front of a condition group
+                - possible string representation:
+                    - "NOT"
+                    - "not"
+                    - "!" (default serialization string representation)
+
+        Different operators cannot be used within the same condition group, parenthesis are necessary in order to explicit operator resolution.
```
Comment (ericbrts) [inline perceval/utils/postselect.py:102]: I'd say:
to explicit resolution order

File: perceval/utils/postselect.py [hunk 8]
```diff
@@ -26,206 +26,190 @@
...
-        return self
-
-    def _add_condition(self, indexes, operator: callable, value: int):
-        indexes = (indexes,) if isinstance(indexes, int) else tuple(indexes)
-        if operator not in self._conditions:
-            self._conditions[operator] = []
-        self._conditions[operator].append((indexes, value))
+    >>> ps = PostSelect("([0,1] == 1 & [2] > 1) | [2] == 0") # Means "I want either exactly one photon in mode 0 or 1, and at least one photon in mode 2, either no photon in mode 2"
+    >>> print(ps(BasicState([0, 1, 1])))
+    True
```
Comment (ericbrts) [inline perceval/utils/postselect.py:117]: Isn't this one False?

File: docs/source/legacy.rst [hunk 9]
```diff
@@ -27,6 +27,24 @@ or
 
 >>> results = simulator.probs_svd(svd, my_detector_list, my_callback)  # my_detector_list can be None
 
+PostSelect API changes
+^^^^^^^^^^^^^^^^^^^^^^
+
+The :code:`PostSelect` class, used to represent a set of post-selection conditions, was moved from Python to C++. This
+change allows to support a richer boolean syntax, with nested condition based on more than the `and` operator.
```
Comment (raksharuia) [inline docs/source/legacy.rst:34]: This change allows supporting a richer boolean syntax, including nested conditions based on operators in addition to `and`.

File: docs/source/legacy.rst [hunk 10]
```diff
@@ -27,6 +27,24 @@ or
...
+
+Removal of operator methods
++++++++++++++++++++++++++++
+
+Now that we support nested logical expression, it made no sense to be able to add conditions one by one. As such,
```
Comment (raksharuia) [inline docs/source/legacy.rst:39]: Now that we support nested logical expression, adding conditions one by one no longer makes sense. Consequently, ...

File: docs/source/legacy.rst [hunk 11]
```diff
@@ -27,6 +27,24 @@ or
...
+:code:`eq`, :code:`gt`, :code:`lt`, :code:`ge` and :code:`le` methods were removed. All :code:`PostSelect` must now be
+constructed from a string or through merging two existing post-selection objects.
+
+In-place apply permutation
+++++++++++++++++++++++++++
+
+Previously, :code:`apply_permutation` method would create a new :code:`PostSelect` object and return it. For
+consistency purpose, the new behavior is to work on the data of the instance :code:`apply_permutation` is called on.
```
Comment (raksharuia) [inline docs/source/legacy.rst:47]: .. purpose, the new behavior *modifies* the data of the instance *on which* ...

File: perceval/utils/postselect.py (hunk 8 above)
Comment (raksharuia) [reply perceval/utils/postselect.py:117]: either is usually followed by or, so I would suggest # Means "I want either exactly one photon in mode 0 or 1, and at least one photon in mode 2, **or** no photon in mode 2"

File: perceval/utils/postselect.py [hunk 12]
```diff
@@ -26,206 +26,190 @@
...
+        super().apply_permutation(perm_vector, first_mode)
 
-    def shift_modes(self, shift: int):
+    def can_compose_with(self, modes: list[int]) -> bool:
         """
-        Shift all mode indexes inside this instance
+        Check if all conditions are compatible with a composition on given modes.
 
-        :param shift: Value to shift all mode indexes with
+        Compatible means that modes list is either a subset, either doesn't intersect with any mode list of any conditions.
```
Comment (raksharuia) [inline perceval/utils/postselect.py:162]: Same comment here, use either..or instead of either..either

File: perceval/utils/postselect.py [hunk 13]
```diff
@@ -26,206 +26,190 @@
...
         bsd: BSDistribution,
         postselect: PostSelect,
         heralds: dict = None,
         keep_heralds: bool = True) -> tuple[BSDistribution, float]:
+    """Post select a BSDistribution
+
+    :param bsd: BSDistribution to post select
+    :param postselect: post selection to apply
+    :param heralds: heralds to apply, defaults to None
+    :param keep_heralds: Keep heralds modes in the BSDistribution (heralds modes will be remove from Basic state), defaults to True
```
Comment (raksharuia) [inline perceval/utils/postselect.py:210]: I think it should be - *heralded* modes will be *removed* ...

File: perceval/utils/postselect.py [hunk 14]
```diff
@@ -254,6 +238,14 @@ def post_select_statevector(
         postselect: PostSelect,
         heralds: dict = None,
         keep_heralds: bool = True) -> tuple[StateVector, float]:
+    """Post select a state vector
+
+    :param sv: State Vector to post select
+    :param postselect: post selection to apply
+    :param heralds: heralds to apply, defaults to None
+    :param keep_heralds: Keep heralds modes in the BSDistribution (heralds modes will be remove from Basic state), defaults to True
```
Comment (raksharuia) [inline perceval/utils/postselect.py:246]: same comment here

**Review by raksharuia — CHANGES_REQUESTED** [review]

**Pull request merged**